File: 프로그래머스/lv3/42628. 이중우선순위큐/이중우선순위큐.py
import heapq

def solution(operations):
    answer = [0, 0]
    max_heap = []
    min_heap = []
    popped = [False for _ in range(len(operations))]
    # heappush: maxheap과 minheap을 만들어서 둘모두에게 push한다 
    # popped를 체크해서 pop된 node인지 아닌지 체크한다
    
    # heappop: 이미 popped된 node이면 계속해서 pop한다
    for i, op in enumerate(operations):
        cmd, num = op.split(" ")
        
        if cmd == 'I':
            heapq.heappush(min_heap, (int(num), i))
            heapq.heappush(max_heap, (int(num) * -1, i))
            
        elif cmd == 'D' and num == '1' and max_heap:
            node, i = heapq.heappop(max_heap)
            while max_heap and popped[i]:
                node, i = heapq.heappop(max_heap)
            popped[i] = True
            
        elif cmd == 'D' and num == '-1' and min_heap:
            node, i = heapq.heappop(min_heap)
            while min_heap and popped[i]:
                node, i = heapq.heappop(min_heap)
            popped[i] = True
    
    # 모든 연산이 끝나고 최소, 최대 값을 저장한다
    if not max_heap or not min_heap:
        return answer
    # 최대값
    node, i = heapq.heappop(max_heap)
    while max_heap and popped[i]:
        node, i = heapq.heappop(max_heap)
    answer[0] = node * -1
    # 최소값
    node, i = heapq.heappop(min_heap)
    while min_heap and popped[i]:
        node, i = heapq.heappop(min_heap)
    answer[1] = node
    
    return answer

File: 프로그래머스/lv3/42628. 이중우선순위큐/이중우선순위큐.java
import java.util.*;

class Solution {
    public int[] solution(String[] operations) {
        int[] answer = {0, 0};
        
        PriorityQueue<Integer> pq = new PriorityQueue<>();
        PriorityQueue<Integer> pq_reverse = new PriorityQueue<>(Collections.reverseOrder());
        
        for (int i = 0; i < operations.length; i++) {
            String[] operation = operations[i].split(" ");
            int num = Integer.parseInt(operation[1]);
            
            if (operation[0].equals("I")) {
                pq.add(num);
                pq_reverse.add(num);
            } else if (num > 0 && pq_reverse.size() > 0) {
                int min = pq_reverse.poll();
                pq.remove(min);
            } else if (pq.size() > 0) {
                int max = pq.poll();
                pq_reverse.remove(max);
            }
        }
        if (pq.size() > 1) {
            answer[1] = pq.poll();
            answer[0] = pq_reverse.poll();
        }
        return answer;
    }
}